### ISSUE TITLE
Name, rendering not on layer 16

### ISSUE THREAD

**Issue by AllroadsNL** [gravitystorm/openstreetmap-carto#2429]
name=Autódromo Santiago "Yaco" Guarnieri
on a polygon, leisure=sports_centre , sport=motor

[Layer 15 it is](http://www.openstreetmap.org/#map=15/-27.3917/-58.9980)
but not on
[Layer 16(http://www.openstreetmap.org/#map=16/-27.3876/-59.0054)
then on
[Layer 17 it is back ](http://www.openstreetmap.org/#map=17/-27.38896/-59.00283)

**Comment by pnorman** [comment]
z17
![image](https://cloud.githubusercontent.com/assets/1190866/20096010/cb5d1c4e-a55c-11e6-8c09-c1fdfce54a8d.png)

Looks like a collision between the icons and text happens on z16.

**Comment by kocio-pl** [comment]
I would like the parking icon ("P") to have lower priority than names in general. Is it possible?

**Comment by meased** [comment]
> I would like the parking icon ("P") to have lower priority than names in general.

:+1: 

Usually when I add parking lots to an area the rendering gets worse.

**Comment by pnorman** [comment]
> I would like the parking icon ("P") to have lower priority than names in general. Is it possible?

Yes. My guess on how to best implement it would to be move parking to the low priority amenity points layer.

**Comment by AllroadsNL** [comment]
Although both should be visible at the layers. Name above the P icon.

**Comment by kocio-pl** [comment]
Unfortunately I'm not too familiar with layers. 

Parking is already included in [amenity-low-priority-poly](https://github.com/gravitystorm/openstreetmap-carto/blob/master/project.yaml#L2499), but probably this layer should be moved earlier (as it was with https://github.com/gravitystorm/openstreetmap-carto/pull/2394)?

**Comment by kocio-pl** [comment]
I've made map.osm export from this area and was trying to change tags in text editor to see the difference. It looks like both oneway arrows and parkings are blocking the name from rendering on z16 - changing both tags allows it to be seen.

I was also trying to remove just the arrows from the data and play with parking layers - but even placing "amenity-points-poly" and "amenity-points" as the first layers was not working. Only removing these layers allows the name to be rendered.

Maybe someone more experienced with layers could help?

**Comment by pnorman** [comment]
> Maybe someone more experienced with layers could help?

Good time to learn ;)

When it comes to placements with collisions, each layer is rendered in the order specified, and collisions bounding boxes are generated. Most symbols won't place where there is an existing symbol. You can see the order of the layers in the Layer array in project.mml, or extract the names with this bit of Python

```python
import yaml
for l in yaml.safe_load(open("project.mml"))["Layer"]:
  print(l["name"])

```

Because amenity=parking is already in the last two layers, moving it won't change anything.

The oneway arrows are probably colliding with the name. I'm not sure what the best solution here is.

cf. #2320 #1840 #1968

**Comment by kocio-pl** [comment]
> Good time to learn ;)

Thanks for your comments and a code snippet!

I think the most useful tool for learning this would be text editor with support for collapsing and moving the whole layers. Looks like Atom can do that, however I don't know yet if I could collapse all for a start and if moving is supported more easily than just copy/paste.

> Because amenity=parking is already in the last two layers, moving it won't change anything.

I did my homework. :smile: I haven't mentioned it, but moving these last two layers ("amenity-low-priority" and "amenity-low-priority-poly") to the top along with "amenity-points-poly" and "amenity-points" did not help, so there's something I still don't understand.

**Comment by pnorman** [comment]
> I think the most useful tool for learning this would be text editor with support for collapsing and moving the whole layers. Looks like Atom can do that, however I don't know yet if I could collapse all for a start and if moving is supported more easily than just copy/paste.

Any decent text editor should have folding capabilities, but you might have to manually tell it the file is YAML

As an example

![image](https://cloud.githubusercontent.com/assets/1190866/21084967/e50608d6-bfbf-11e6-808e-ba54cd34ba32.png)

If I position my cursor at the start of the array element with `id: waterway_name` and press shift-down arrow it will select the entire element, I cut it, and I can then put my cursor at the start of the element with `id: building`, and paste it in before that layer.

Back to this problem, I can see a few ways to fix this

1. Make oneway arrows non-colliding
2. Move road names + oneway arrows later
3. Move oneway arrows out of the road name layer and put them into their own layer, positioned later.
4. Radically change the oneway styling

I don't think 2 will work since we might end up with no road names in POI-dense areas.

1 would be easy. 3 makes our roads code bigger :( 4 requires ideas.

I'm concerned about the size of our roads code, and really wouldn't like to see it grow any larger. I'd be inclined towards 1, which is how it used to be.

**Comment by kocio-pl** [comment]
There's a problem that folding happens at "-", so with two layers (`- name: "roads-casing"` and `- name: "admin-low-zoom"`) it tries to fold from name, leaving id in the previous layer. This would of course not hurt this case, since we don't need to touch these layers, but could we fix it without causing other problems?

I was thinking about simply dragging folded elements as the most convenient way, but I can live without it. Also folding all elements at the start would be great and maybe [fold-functions plugin](https://atom.io/packages/fold-functions) can do it, but I've just found that Atom remembers folding state of the code, so it's also not a big deal for me now.

**Comment by pnorman** [comment]
Folding still works fine on those layers, folding the entire array element into one line.

The following YAML is the same, just with a different text representation

```yaml
  - name: "roads-casing"
    id: "roads-casing"
    geometry: "linestring"
```
```yaml
  - id: "roads-casing"
    name: "roads-casing"
    geometry: "linestring"
```
```yaml
  - 
    id: "roads-casing"
    name: "roads-casing"
    geometry: "linestring"
```

```yaml
- {geometry: linestring, id: roads-casing, name: roads-casing}
```

I'm not aware of any editors which allow re-ordering of a YAML array by drag-and-drop, mainly because I don't know of any *YAML* editors, just text editors. I am aware of JSON and XML editors, but normally gets used in situations where you don't need an editor for it and can edit the raw text.

**Comment by kocio-pl** [comment]
Thanks. Just double checking: so I guess I can make a patch moving both "-" (dashes) from `name:` to corresponding `id:`, like it's done in the rest of layers? Currently we have:

```yaml
    id: "roads-casing"
  - name: "roads-casing"
    geometry: "linestring"
```

which is of course not the same as any of your proper examples, like:

```yaml
  - name: "roads-casing"
    id: "roads-casing"
    geometry: "linestring"
```

and that's why folding is broken at the moment.

**Comment by pnorman** [comment]
> Currently we have:

No we don't.

![image](https://cloud.githubusercontent.com/assets/1190866/21086618/fde45246-bfd2-11e6-99d5-42e9939e5fc5.png)

The meaning of the two is completely different. `-` starts an array element, so in the first you pasted the array element with `name: roads-casing` does not have `id: roads-casing`

**Comment by kocio-pl** [comment]
> No we don't.

I had it in my code and I don't know how did it happen, but that's not important now - you're right. Still a cosmetic patch making things uniform would make me happy.

**Comment by kocio-pl** [comment]
> I'm concerned about the size of our roads code, and really wouldn't like to see it grow any larger. I'd be inclined towards 1, which is how it used to be.

Why was it changed? I can't find the right PR. I though we rather [made arrows less visible than names](https://github.com/gravitystorm/openstreetmap-carto/pull/2320)?

I have no favorite solution, but any would be welcome. No 1 reason I see is that universal style is rather to show names than help routing - they are useful hint, so I have nothing against showing them, but we have specialized software for routing.

**Comment by pnorman** [comment]
> Why was it changed?

We changed to arrows made from dashed lines to SVGs. #1840

**Comment by kocio-pl** [comment]
I haven't notice that it was also layers change. Making it a low priority layer again would be good enough for me.

**Comment by pnorman** [comment]
I consider showing arrows important.

> Making it a low priority layer again

It wasn't a low priority layer before. When it was dashes, it didn't conflict with any other placements.

**Comment by dieterdreist** [comment]
sent from a phone

> On 14 Dec 2016, at 17:44, Paul Norman <notifications@github.com> wrote:
> 
> It wasn't a low priority layer before. When it was dashes, it didn't conflict with any other placements.


is there a technical or cartographic reason we don't want to allow overlap of later rendered stuff with them?

**Comment by math1985** [comment]
There is no technical reason.

**Comment by kocio-pl** [comment]
Looks like a duplicate of https://github.com/gravitystorm/openstreetmap-carto/issues/964.

**Comment by pnorman** [comment]
> Looks like a duplicate of #964.

This is specific to oneway arrows, which are not POI labels and can have a different solution.

**Comment by jeisenbe** [comment]
Anyone still interested in trying "1. Make oneway arrows non-colliding" again?

**Comment by imagico** [comment]
#4192 showed this is also a problem with point symbol labels.
